File: src/analyzer/declare/declare-implicit-arguments.ts
import type { Scope } from "../../ir/scope/scope.js";
import { VariableImpl } from "../variable-impl.js";

// eslint-scope's FunctionScope.__defineArguments registers an `arguments`
// Variable with no identifiers and no defs (CreateUnmappedArgumentsObject /
// CreateMappedArgumentsObject from FunctionDeclarationInstantiation, ES spec
// 9.2.13). Arrow functions inherit `arguments` from the enclosing scope and
// must not call this helper.
export function declareImplicitArguments(scope: Scope): void {
  if (scope.set.has("arguments")) {
    return;
  }
  const variable = new VariableImpl("arguments", scope);
  scope.set.set("arguments", variable);
  scope.variables.push(variable);
}

File: crates/unsnarl-boundary-eslint-scope/src/declare/declare_implicit_arguments.rs
//! Register the synthetic `arguments` binding inside a function scope.
//!
//! Pushes a fresh `VariableData` with no identifiers and no defs —
//! the ES spec's `CreateUnmappedArgumentsObject` /
//! `CreateMappedArgumentsObject` shape. Arrow functions inherit
//! `arguments` from the enclosing function scope and must not call
//! this helper.

use unsnarl_ir::ids::ScopeId;
use unsnarl_ir::scope::VariableData;

use crate::state::ScopeBuilderState;

pub(crate) fn declare_implicit_arguments(state: &mut ScopeBuilderState, scope: ScopeId) {
    let name = "arguments";
    if state.arena.scopes[scope].set().contains_key(name) {
        return;
    }
    let id = state.arena.variables.push(VariableData::new(
        name.to_string(),
        scope,
        Vec::new(),
        Vec::new(),
        Vec::new(),
    ));
    state.arena.scopes[scope].insert_into_set(name.to_string(), id);
    state.arena.scopes[scope].variables.push(id);
}

#[cfg(test)]
#[path = "declare_implicit_arguments_test.rs"]
mod declare_implicit_arguments_test;
